src/client/datascience/baseJupyterSession.ts

```diff
@@ -377,6 +377,7 @@ export abstract class BaseJupyterSession implements IJupyterSession {
                 if (!this.canShutdownSession(session, isRequestToShutDownRestartSession, shutdownEvenIfRemote)) {
                     traceVerbose(`Session cannot be shutdown ${session.kernelConnectionMetadata?.id}`);
                     session.dispose();
+                    this._disposed.fire();
```

Comment:
I don't think this is correct.
In the old class `disposed` is fired in only one place thats the `shutdownImplementation`
This method is used only to shutdown a specific session object, not to shutdown (dispose) this entire class.

Looking at some code in other places I can see that's what's expected
I.e. `this._disposed.fire` should only be called when the entire session is killed off (either via a dispose or explicit call to the `shutdown` method)
```
            // Shutdown the session but not this class.
            this.setSession(undefined);
            this.shutdownSession(session, this.statusHandler, false)
```

Reply:
I'd like to remove this and clean up so that we have the disposed called from just one place. So we know when and where the class is disposed.